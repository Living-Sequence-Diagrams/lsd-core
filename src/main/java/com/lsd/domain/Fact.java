package com.lsd.domain;

import lombok.Builder;
import lombok.Value;

@Value
@Builder
public class Fact {
    String key;
    String value;
}
