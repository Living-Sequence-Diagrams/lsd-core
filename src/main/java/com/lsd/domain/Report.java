package com.lsd.domain;

import lombok.Builder;
import lombok.Value;

import java.util.Collection;

@Value
@Builder
public class Report {
    String title;
    Collection<Scenario> scenarios;
}
