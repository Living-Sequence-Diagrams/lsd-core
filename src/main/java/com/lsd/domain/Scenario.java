package com.lsd.domain;

import com.lsd.PlantUmlRenderer;
import com.lsd.SvgConverter;
import com.lsd.domain.interactions.Interaction;
import lombok.Builder;
import lombok.Value;

import java.util.Collection;

@Value
@Builder
public class Scenario {
    String title;
    String description;
    Collection<Interaction> interactions;
    Collection<Fact> facts;

    PlantUmlRenderer plantUmlRenderer = new PlantUmlRenderer();
    SvgConverter svgConverter = new SvgConverter();

    public String sequenceDiagram() {
        String plantUml = plantUmlRenderer.render(interactions);
        return svgConverter.convert(plantUml);
    }
}
