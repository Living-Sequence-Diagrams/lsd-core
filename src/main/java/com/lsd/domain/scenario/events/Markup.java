package com.lsd.domain.scenario.events;

import lombok.RequiredArgsConstructor;
import lombok.Value;

@Value
@RequiredArgsConstructor
public class Markup implements Event {
    String value;

    public String toMarkup() {
        return value;
    }
}
