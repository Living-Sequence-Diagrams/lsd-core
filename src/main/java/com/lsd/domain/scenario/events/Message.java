package com.lsd.domain.scenario.events;

import lombok.Builder;
import lombok.Value;

import static java.lang.String.format;

@Value
@Builder
public class Message implements Interaction {
    String id;
    String label;
    String from;
    String to;
    Object data;

    public String toMarkup() {
        return format("%s -> %s:[[#%s %s]]",
                getFrom(), getTo(), getId(), getLabel());
    }
}
