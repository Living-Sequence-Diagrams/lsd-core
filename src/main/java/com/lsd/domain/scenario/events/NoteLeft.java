package com.lsd.domain.scenario.events;

import lombok.RequiredArgsConstructor;
import lombok.Value;

import static java.lang.String.format;

@Value
@RequiredArgsConstructor
public class NoteLeft implements Event {
    String value;

    public String toMarkup() {
        return format("note left: %s", getValue());
    }
}
