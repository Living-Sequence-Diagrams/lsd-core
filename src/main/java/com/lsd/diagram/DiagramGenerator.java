package com.lsd.diagram;

import com.lsd.properties.LsdProperties;
import com.lsd.report.model.Participant;
import com.lsd.report.model.SequenceDiagram;
import com.lsd.events.SequenceEvent;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.toList;

@Builder
public class DiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<SequenceEvent> events;

    public SequenceDiagram sequenceDiagram() {
        String uml = generateSequenceUml();
        String svg = generateSequenceSvg(uml);
        return SequenceDiagram.builder()
                .uml(uml)
                .svg(svg)
                .build();
    }

    private String generateSequenceSvg(String markup) {
        SvgConverter svgConverter = new SvgConverter();

        return svgConverter.convert(markup);
    }

    @SneakyThrows
    private String generateSequenceUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants,
                "events", events.stream()
                        .map(SequenceEvent::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
