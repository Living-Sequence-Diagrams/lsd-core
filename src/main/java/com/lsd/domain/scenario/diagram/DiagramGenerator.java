package com.lsd.domain.scenario.diagram;

import com.lsd.domain.Participant;
import com.lsd.domain.scenario.events.Event;
import com.lsd.domain.scenario.events.Interaction;
import com.lsd.properties.LsdProperties;
import com.mitchellbosecke.pebble.PebbleEngine;
import com.mitchellbosecke.pebble.template.PebbleTemplate;
import lombok.Builder;
import lombok.SneakyThrows;

import java.io.StringWriter;
import java.io.Writer;
import java.util.List;
import java.util.Map;
import java.util.Set;

import static com.lsd.properties.LsdProperties.DIAGRAM_THEME;
import static java.util.stream.Collectors.toList;

@Builder
public class DiagramGenerator {
    private final PebbleEngine engine = new PebbleEngine.Builder().autoEscaping(false).build();
    private final PebbleTemplate compiledTemplate = engine.getTemplate("templates/sequence-uml.peb");

    private final Set<String> includes;
    private final List<Participant> participants;
    private final List<Event> events;

    public SequenceDiagram sequenceDiagram() {
        return SequenceDiagram.builder()
                .uml(generateSequenceUml())
                .interactions(events.stream()
                        .filter(Interaction.class::isInstance)
                        .map(Interaction.class::cast)
                        .collect(toList()))
                .build();
    }

    @SneakyThrows
    private String generateSequenceUml() {
        Writer writer = new StringWriter();
        compiledTemplate.evaluate(writer, Map.of(
                "theme", LsdProperties.get(DIAGRAM_THEME),
                "includes", includes,
                "participants", participants,
                "events", events.stream()
                        .map(Event::toMarkup)
                        .collect(toList())
        ));
        return writer.toString();
    }
}
