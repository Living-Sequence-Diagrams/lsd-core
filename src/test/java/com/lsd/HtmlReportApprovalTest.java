package com.lsd;

import com.lsd.domain.scenario.Fact;
import com.lsd.domain.Report;
import com.lsd.domain.scenario.Scenario;
import com.lsd.domain.scenario.diagram.DiagramGenerator;
import com.lsd.domain.scenario.events.Message;
import com.lsd.domain.scenario.events.NoteLeft;
import com.lsd.domain.scenario.events.Markup;
import com.lsd.domain.scenario.events.ResponseMessage;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Set;

import static com.lsd.domain.ParticipantType.*;

class HtmlReportApprovalTest {
    private final HtmlReportRenderer htmlReportRenderer = new HtmlReportRenderer();

    @Test
    void renderHtml() {
        String report = htmlReportRenderer.render(Report.builder()
                .title("LSD")
                .scenarios(List.of(
                        Scenario.builder()
                                .title("First Scenario")
                                .description("<pre>Given a Cat\nAnd a dog\nWhen the dog barks\nThen the cat hisses</pre>")
                                .facts(List.of(
                                        Fact.builder()
                                                .key("Greeting")
                                                .value("Hi")
                                                .build())
                                )
                                .sequenceDiagram(DiagramGenerator.builder()
                                        .includes(Set.of())
                                        .participants(List.of(
                                                ACTOR.called("A", "Arnie"),
                                                BOUNDARY.called("Unused participant"),
                                                BOUNDARY.called("Another Unused participant", "G"),
                                                DATABASE.called("B", "Badboy \\nDB")
                                        ))
                                        .events(List.of(
                                                Message.builder()
                                                        .id("111")
                                                        .label("Sending a greeting")
                                                        .from("A")
                                                        .to("B")
                                                        .data("Hi")
                                                        .build(),
                                                new Markup("..."),
                                                ResponseMessage.builder()
                                                        .id("222")
                                                        .label("Sending a response")
                                                        .from("B")
                                                        .to("A")
                                                        .data("Yo")
                                                        .build(),
                                                new NoteLeft("Informal")
                                        ))
                                        .build()
                                        .sequenceDiagram())
                                .build(),
                        Scenario.builder()
                                .title("Second Scenario")
                                .description("<p><ul><li>Some details</li><li>Some more details</li></ul></p>")
                                .sequenceDiagram(DiagramGenerator.builder()
                                        .includes(Set.of())
                                        .participants(List.of())
                                        .events(List.of(
                                                Message.builder()
                                                        .id("333")
                                                        .label("An interaction")
                                                        .from("Alpha")
                                                        .to("Beta")
                                                        .data("{\"name\": \"alpha\", \"description\":\"Something long to see how the popup window will respond to text that is wider than the initial box size\"}")
                                                        .build(),
                                                Message.builder()
                                                        .id("444")
                                                        .label("An interaction")
                                                        .from("Beta")
                                                        .to("Gamma")
                                                        .data("β")
                                                        .build(),
                                                Message.builder()
                                                        .id("555")
                                                        .label("An interaction")
                                                        .from("Gamma")
                                                        .to("Delta")
                                                        .data("γ")
                                                        .build()))
                                        .build().sequenceDiagram())
                                .build()))
                .build());

        Approvals.verifyHtml(report);
    }
}