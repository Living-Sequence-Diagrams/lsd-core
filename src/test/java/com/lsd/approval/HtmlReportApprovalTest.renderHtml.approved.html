<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en">
<head>
    <title>LSD</title>
    <style>
        .overlay {
            position: fixed;
            top: 0;
            bottom: 0;
            left: 0;
            right: 0;
            background: rgba(0, 0, 0, 0.7);
            transition: opacity 500ms;
            visibility: hidden;
            opacity: 0;
        }

        .overlay:target {
            visibility: visible;
            opacity: 1;
        }

        .popup {
            margin: 70px auto;
            padding: 20px;
            background: #fff;
            border-radius: 5px;
            max-width: 70%;
            position: relative;
            transition: all 5s ease-in-out;
        }

        h1, h2, h3 {
            margin-top: 0;
            color: #333;
            font-family: Tahoma, Arial, sans-serif;
        }

        .popup .close {
            position: absolute;
            top: 20px;
            right: 30px;
            transition: all 200ms;
            font-size: 30px;
            font-weight: bold;
            text-decoration: none;
            color: #333;
        }

        .popup .close:hover {
            color: #06D85F;
        }

        .popup .content {
            overflow: auto;
        }

        /*    Make Facts look like table*/
        .factTable {
            display: table;
        }

        .factRow {
            display: table-row;
        }

        .factRowHead {
            display: table-header-group;
            font-weight: bold;
        }

        .factCell, .factRowHead {
            display: table-cell;
            padding: 3px 10px;
            border: 1px solid #999999;
        }

        /* Some basic spacing for now  */
        section {
            border: 15px solid skyblue;
            padding: 50px;
            margin: 20px;
        }
    </style>
</head>
<body>
<main role="main">
    <h1>LSD</h1>
        <article>
        <h2>First Scenario</h2>
        <section class="description">
            <h3>Description</h3>
            <pre>Given a Cat
And a dog
When the dog barks
Then the cat hisses</pre>        </section>
                <section class="facts">
            <h3>Key Facts</h3>
            <div class="factTable">
                <div class="factRow">
                    <div class="factRowHead">Key</div>
                    <div class="factRowHead">Value</div>
                </div>
                                <div class="factRow">
                    <div class="factCell">Greeting</div>
                    <div class="factCell">Hi</div>
                </div>
                            </div>
        </section>
                <section class="sequence diagram svg">
            <h3>Sequence Diagram</h3>
            <?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentScriptType="application/ecmascript" contentStyleType="text/css" height="301px" preserveAspectRatio="none" style="width:297px;height:301px;background:#FFFFFF;" version="1.1" viewBox="0 0 297 301" width="297px" zoomAndPan="magnify">
  <defs />
  <g>
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="107" x2="107" y1="85.0293" y2="128.2293" />
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:1.0,4.0;" x1="107" x2="107" y1="128.2293" y2="156.2293" />
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="107" x2="107" y1="156.2293" y2="205.4293" />
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="262" x2="262" y1="85.0293" y2="128.2293" />
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:1.0,4.0;" x1="262" x2="262" y1="128.2293" y2="156.2293" />
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="262" x2="262" y1="156.2293" y2="205.4293" />
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="37" x="86" y="82.0898">Arnie</text>
    <ellipse cx="107.5" cy="16.0146" fill="#FFFFFF" rx="8" ry="8" style="stroke:#000000;stroke-width:1.0;" />
    <path d="M107.5,24.0146 L107.5,51.0146 M94.5,32.0146 L120.5,32.0146 M107.5,51.0146 L94.5,66.0146 M107.5,51.0146 L120.5,66.0146 " fill="none" style="stroke:#000000;stroke-width:1.0;" />
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="37" x="86" y="218.5045">Arnie</text>
    <ellipse cx="107.5" cy="230.4439" fill="#FFFFFF" rx="8" ry="8" style="stroke:#000000;stroke-width:1.0;" />
    <path d="M107.5,238.4439 L107.5,265.4439 M94.5,246.4439 L120.5,246.4439 M107.5,265.4439 L94.5,280.4439 M107.5,265.4439 L120.5,280.4439 " fill="none" style="stroke:#000000;stroke-width:1.0;" />
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="52" x="231" y="65.0752">Badboy</text>
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="21" x="249" y="82.0898">DB</text>
    <path d="M244.5,15 C244.5,5 262.5,5 262.5,5 C262.5,5 280.5,5 280.5,15 L280.5,41 C280.5,51 262.5,51 262.5,51 C262.5,51 244.5,51 244.5,41 L244.5,15 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.5;" />
    <path d="M244.5,15 C244.5,25 262.5,25 262.5,25 C262.5,25 280.5,25 280.5,15 " fill="none" style="stroke:#000000;stroke-width:1.5;" />
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="52" x="231" y="218.5045">Badboy</text>
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="21" x="249" y="235.5191">DB</text>
    <path d="M244.5,248.4586 C244.5,238.4586 262.5,238.4586 262.5,238.4586 C262.5,238.4586 280.5,238.4586 280.5,248.4586 L280.5,274.4586 C280.5,284.4586 262.5,284.4586 262.5,284.4586 C262.5,284.4586 244.5,284.4586 244.5,274.4586 L244.5,248.4586 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.5;" />
    <path d="M244.5,248.4586 C244.5,258.4586 262.5,258.4586 262.5,258.4586 C262.5,258.4586 280.5,258.4586 280.5,248.4586 " fill="none" style="stroke:#000000;stroke-width:1.5;" />
    <polygon fill="#000000" points="250.5,116.2293,260.5,120.2293,250.5,124.2293,254.5,120.2293" style="stroke:#000000;stroke-width:1.0;" />
    <line style="stroke:#000000;stroke-width:1.0;" x1="107.5" x2="256.5" y1="120.2293" y2="120.2293" />
    <a href="#111" target="_top" title="#111" xlink:actuate="onRequest" xlink:href="#111" xlink:show="new" xlink:title="#111" xlink:type="simple">
      <text fill="#0000FF" font-family="Verdana" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="87" x="114.5" y="112.0991">Sending food</text>
    </a>
    <a href="#111" target="_top" title="#111" xlink:actuate="onRequest" xlink:href="#111" xlink:show="new" xlink:title="#111" xlink:type="simple">
      <image height="19" width="19" x="206.5" xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABMAAAATCAIAAAD9MqGbAAAA9ElEQVR4Xp2TMQqDMBSGU0S6dquCnQrewsnByYMIbt6idCgUurk7uXiH3qHgIoJ0KCiCYKXwmlSD6SuE1J9v0OT/4hsigaUheEFI3+MVMdh8PMDzYLuF1QoIYRgGWylLVBTMuobdbmqPOM7Xq22zc7F5u4GmsW1RjuP5ebNhBV2HPBfMppk1sS1yOMB+Pz13HTejCPfk+D43kwTvyTkeuUlzvUKWKUGbY5h5v8P5zE5S4XKBquLm7zxy1mt4vRaZ9JK07cek01sWmKYqacqnHQYoij94PrkZBHgeOa7LzdMJ78kJQ27Sr6O7LoE2x+C/TD3LzTeit8koBlhA2AAAAABJRU5ErkJggg==" y="99.0293" />
    </a>
    <line style="stroke:#000000;stroke-width:1.0;" x1="107.5" x2="117.5" y1="182.729" y2="178.729" />
    <line style="stroke:#000000;stroke-width:1.0;" x1="107.5" x2="117.5" y1="182.729" y2="186.729" />
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:2.0,2.0;" x1="107.5" x2="261.5" y1="182.729" y2="182.729" />
    <a href="#222" target="_top" title="#222" xlink:actuate="onRequest" xlink:href="#222" xlink:show="new" xlink:title="#222" xlink:type="simple">
      <text fill="#0000FF" font-family="Verdana" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="131" x="124.5" y="177.9995">Sending a response</text>
    </a>
    <path d="M10,161.2293 L10,190.2293 L103,190.2293 L103,171.2293 L93,161.2293 L10,161.2293 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;" />
    <path d="M93,161.2293 L93,171.2293 L103,171.2293 L93,161.2293 " fill="#FFFFFF" style="stroke:#000000;stroke-width:1.0;" />
    <text fill="#000000" font-family="Verdana" font-size="13" lengthAdjust="spacing" textLength="48" x="16" y="182.6998">Friends</text>
    <image height="19" width="19" x="69" xlink:href="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAABMAAAATCAIAAAD9MqGbAAAA+0lEQVR4Xp2Sz6pFUByFtygzGRgpD+A5lLGHUF7C1NhAKRkbGCkPwAN4C4bGQuTenz/XOVbHPeWbWWt9drTZz1MYPE/TBMnOMAyQvEzDMARBYIxJkuQ4zpnbtk0J5aIomqY5z/Oer2bf97IssyuKonRdp6oq5MQur6ZlWVhu8DyP0QadfJjYfIPeuCzLE5No2/ahOY7jamqahs2/6Lp+fGdd13c/4yNN0xwmUVUV9jeUZbkrr5tQFAXHcTi8kqbpub/cvjzPcftGlmXvY7y3SZKgsRGGISzRJOI4Bi2KIhx9NAnf90/NdV2sNz6bhOd5pAVBgMUft+ZXnpu/Yy0pOXekRvwAAAAASUVORK5CYII=" y="166.2293" />
  </g>
</svg>

        </section>
        <section class="sequence diagram uml">
            <h3>Sequence Diagram Markup</h3>
            <pre>@startuml

!theme plain
hide unlinked

!include &lt;tupadr3/font-awesome-5/hamburger&gt;
!include &lt;tupadr3/font-awesome-5/heart&gt;

actor A as &quot;Arnie&quot;
boundary UnusedParticipant
boundary AnotherUnusedParticipant as &quot;G&quot;
database B as &quot;Badboy \nDB&quot;

A -&gt; B:[[#111 Sending food &lt;$hamburger{scale=0.4}&gt;]]
...
B --&gt;&gt; A:[[#222 Sending a response]]
note left: Friends &lt;$heart{scale=0.4}&gt;

@enduml</pre>
        </section>
                <!-- Section containing data for sequence diagram popups -->
        <div id="111" class="overlay">
            <div class="popup">
                <h2>Sending food </h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>Hi</pre>
                </div>
            </div>
        </div>
                <!-- Section containing data for sequence diagram popups -->
        <div id="222" class="overlay">
            <div class="popup">
                <h2>Sending a response</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>Thank You!</pre>
                </div>
            </div>
        </div>
            </article>
        <article>
        <h2>Second Scenario</h2>
        <section class="description">
            <h3>Description</h3>
            <p><ul><li>Some details</li><li>Some more details</li></ul></p>        </section>
                <section class="sequence diagram svg">
            <h3>Sequence Diagram</h3>
            <?xml version="1.0" encoding="UTF-8"?>
<svg xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink" contentScriptType="application/ecmascript" contentStyleType="text/css" height="182px" preserveAspectRatio="none" style="width:660px;height:182px;background:#FFFFFF;" version="1.1" viewBox="0 0 660 182" width="660px" zoomAndPan="magnify">
  <defs />
  <g>
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="47" x2="47" y1="37.0146" y2="146.4126" />
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="164" x2="164" y1="37.0146" y2="146.4126" />
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="511.5" x2="511.5" y1="37.0146" y2="146.4126" />
    <line style="stroke:#000000;stroke-width:1.0;stroke-dasharray:5.0,5.0;" x1="628.5" x2="628.5" y1="37.0146" y2="146.4126" />
    <rect fill="#FFFFFF" height="31.0146" style="stroke:#000000;stroke-width:1.0;" width="54" x="20" y="5" />
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="40" x="27" y="26.0752">Alpha</text>
    <rect fill="#FFFFFF" height="31.0146" style="stroke:#000000;stroke-width:1.0;" width="54" x="20" y="145.4126" />
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="40" x="27" y="166.4878">Alpha</text>
    <rect fill="#FFFFFF" height="31.0146" style="stroke:#000000;stroke-width:1.0;" width="46" x="141" y="5" />
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="32" x="148" y="26.0752">Beta</text>
    <rect fill="#FFFFFF" height="31.0146" style="stroke:#000000;stroke-width:1.0;" width="46" x="141" y="145.4126" />
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="32" x="148" y="166.4878">Beta</text>
    <rect fill="#FFFFFF" height="31.0146" style="stroke:#000000;stroke-width:1.0;" width="69" x="477.5" y="5" />
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="55" x="484.5" y="26.0752">Gamma</text>
    <rect fill="#FFFFFF" height="31.0146" style="stroke:#000000;stroke-width:1.0;" width="69" x="477.5" y="145.4126" />
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="55" x="484.5" y="166.4878">Gamma</text>
    <rect fill="#FFFFFF" height="31.0146" style="stroke:#000000;stroke-width:1.0;" width="51" x="603.5" y="5" />
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="37" x="610.5" y="26.0752">Delta</text>
    <rect fill="#FFFFFF" height="31.0146" style="stroke:#000000;stroke-width:1.0;" width="51" x="603.5" y="145.4126" />
    <text fill="#000000" font-family="Verdana" font-size="14" lengthAdjust="spacing" textLength="37" x="610.5" y="166.4878">Delta</text>
    <polygon fill="#000000" points="152,64.814,162,68.814,152,72.814,156,68.814" style="stroke:#000000;stroke-width:1.0;" />
    <line style="stroke:#000000;stroke-width:1.0;" x1="47" x2="158" y1="68.814" y2="68.814" />
    <a href="#333" target="_top" title="#333" xlink:actuate="onRequest" xlink:href="#333" xlink:show="new" xlink:title="#333" xlink:type="simple">
      <text fill="#0000FF" font-family="Verdana" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="93" x="54" y="64.0845">An interaction</text>
    </a>
    <polygon fill="#000000" points="500,94.6133,510,98.6133,500,102.6133,504,98.6133" style="stroke:#000000;stroke-width:1.0;" />
    <line style="stroke:#000000;stroke-width:1.0;" x1="164" x2="506" y1="98.6133" y2="98.6133" />
    <a href="#444" target="_top" title="#444" xlink:actuate="onRequest" xlink:href="#444" xlink:show="new" xlink:title="#444" xlink:type="simple">
      <text fill="#0000FF" font-family="Verdana" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="324" x="171" y="93.8838">An interaction description that is long enough ...</text>
    </a>
    <polygon fill="#000000" points="617,124.4126,627,128.4126,617,132.4126,621,128.4126" style="stroke:#000000;stroke-width:1.0;" />
    <line style="stroke:#000000;stroke-width:1.0;" x1="512" x2="623" y1="128.4126" y2="128.4126" />
    <a href="#555" target="_top" title="#555" xlink:actuate="onRequest" xlink:href="#555" xlink:show="new" xlink:title="#555" xlink:type="simple">
      <text fill="#0000FF" font-family="Verdana" font-size="13" lengthAdjust="spacing" text-decoration="underline" textLength="93" x="519" y="123.6831">An interaction</text>
    </a>
  </g>
</svg>

        </section>
        <section class="sequence diagram uml">
            <h3>Sequence Diagram Markup</h3>
            <pre>@startuml

!theme plain
hide unlinked



Alpha -&gt; Beta:[[#333 An interaction]]
Beta -&gt; Gamma:[[#444 An interaction description that is long enough ...]]
Gamma -&gt; Delta:[[#555 An interaction]]

@enduml</pre>
        </section>
                <!-- Section containing data for sequence diagram popups -->
        <div id="333" class="overlay">
            <div class="popup">
                <h2>An interaction</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>{"name": "alpha", "description":"Something long to see how the popup window will respond to text that is wider than the initial box size"}</pre>
                </div>
            </div>
        </div>
                <!-- Section containing data for sequence diagram popups -->
        <div id="444" class="overlay">
            <div class="popup">
                <h2>An interaction description that is long enough ...</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>β</pre>
                </div>
            </div>
        </div>
                <!-- Section containing data for sequence diagram popups -->
        <div id="555" class="overlay">
            <div class="popup">
                <h2>An interaction</h2>
                <a class="close" href="#!">&times;</a>
                <div class="content">
                    <pre>γ</pre>
                </div>
            </div>
        </div>
            </article>
    </main>
</body>
</html>
